### PR TITLE
Add taskRules flag to NodeExtension

### PR DESCRIPTION
**Pull request by nhoughto**

### CODE REVIEW

File: src/main/groovy/com/moowork/gradle/node/NodePlugin.groovy [hunk 1]
```diff
@@ -36,10 +36,13 @@ class NodePlugin
 
         addGlobalTypes()
         addTasks()
-        addNpmRule()
-        addYarnRule()
 
         this.project.afterEvaluate {
+            if (this.config.taskRules) {
```
Comment (nhoughto) [inline src/main/groovy/com/moowork/gradle/node/NodePlugin.groovy:41]: not sure about this bit?

File: src/main/groovy/com/moowork/gradle/node/NodeExtension.groovy [hunk 2]
```diff
@@ -33,6 +33,8 @@ class NodeExtension
 
     boolean download = false
 
+    boolean taskRules = true
```
Comment (deepy) [inline src/main/groovy/com/moowork/gradle/node/NodeExtension.groovy:36]: I think we might need a more descriptive name for this, but right now I can't think of a good one

Comment (nhoughto) [reply src/main/groovy/com/moowork/gradle/node/NodeExtension.groovy:36]: i went back and forth about it, it is a gradle plugin so in theory shouldn't be abstracting gradle concepts, task rules is the concept being configured. Could be 'dynamic task generation' or something but that is re-labelling something that already has a name.. a task rule 🤷‍♂️

Comment (nhoughto) [reply src/main/groovy/com/moowork/gradle/node/NodeExtension.groovy:36]: added some doco to solve for descriptiveness =)

**Comment by bsautel** [comment]
Thanks for this PR. We will probably have to port it to the master branch because we did not plan to release new features in the 2.x version.
I am wondering whether the default value should not be false. Indeed, if by default it is true, why is this option configurable?

**Comment by nhoughto** [comment]
i guess i could make it `disableTaskRules`? i was really just following the pattern set by `download`

The reason you want it configurable is so you can turn it off? If you make it off by default that would change the contract for existing users of the plugin.. would break heaps of stuff?

**Comment by bsautel** [comment]
I personally prefer `disableTaskRules` even if you are right, this is not consistent with `download`.

That's right, turning this option to false by default would break backward compatibility. But the version 3.0 already broke compatibility and it is still possible to integrate this change before its release. It's the reason why I think that we could afford to make this change.

I am proposing that since if this feature is enable by default, you just have not to use it and that's it. What do you think about that?
As I said, I don't use this feature and I would be ok to deprecate it, it's maybe the reason why I suggest to disable it by default. 😆

**Comment by deepy** [comment]
I like the name `disableTaskRules` better, it's more obvious what happens.

We could deprecate it in 3 and see if anyone complains about it, but it does add to the out-of-the-box experience in that you can just apply the plugin and get going immediately.

File: docs/node.md [hunk 3]
```diff
@@ -204,6 +204,12 @@ node {
   // If false, it will try to use globally installed node.
   download = true
 
+  //if false (the default), it will enable Gradle Task Rules for `yarn_*`, `npm_*` and `npm_*` task names,
+  //to support automatic task creation.
+  //if true, it will not register task rules when the plugin is applied, 
+  //so all rule-generated tasks will need to be configured manually (excluding static `yarn_setup`,`node_setup`,`npm_install`,`yarn_install`)
```
Comment (deepy) [inline docs/node.md:210]: the static tasks don't use `snake_case`, they're all `camelCase`. e.g. `yarnInstall`, `npmInstall`, `nodeSetup`

**Comment by deepy** [comment]
This looks good to me, I'll test it tonight. I think we can do an additional `2.x` release but with 3.x just around the corner that's the main focus.

I've been busy moving to a larger apartment and I've finally got enough room for the computer and gotten enough boxes away so that I can actually reach it. There's some proxy changes I've gotta finish first, but after that I can have a look at porting this to 3 so we can get it into the final RC

**Comment by deepy** [comment]
With this I can't configure tasks that are created from rules, e.g. the following does not work:
```tasks.named('npm_help').configure { doLast { println "hello" } }```

So we can't merge this one into 2.x as it breaks backwards compatibility, we should still be able to get it into 3.x however.

**Pull request closed without merging**